### PR TITLE
Implement re-connect flow using the Mkt App Oauth as provider

### LINKED ISSUE #3338: Implement re-connect flow using the Mkt App Oauth as provider

### PR DESCRIPTION
**Pull request by diegocurbelo**
Fixes #3338

## Changes proposed in this Pull Request:

This PR adds support for the Stripe App OAuth flow for account re-authentication.

If an account is already connected to a different platform account, it can not complete the normal Connect OAuth flow, the account can't be selected in the Stripe Account selection page... and it shows the following message: `This account is connected to another platform. Please create a new account.`. 
Screenshot 2024-08-08 at 3 19 09 AM

For merchants in this situation, we use the Stripe App OAuth flow for authentication:
Screenshot 2024-08-06 at 11 08 04 PM

## Testing instructions

1. Navigate to WP-Admin > WooCommerce > Settings > Payments > Stripe
    > /wp-admin/admin.php?page=wc-settings&tab=checkout&section=stripe
2. Connect a test account (if not connected already)
3. Click on `Configure connection` and mouse over the `Create or connect a test account button` (or click it), check that the URL is `https://connect.stripe.com`
    Screenshot_2024-08-05_at_2_15_19 AM
4. To test the Stripe App OAuth flow, we need to configure the plugin to use keys from an account connected to another platform. If you don't have an account connected to a 3rd party platform, ping me in slack.
5. Reload the settings page, and check that the `accountId` was updated (1), and mouse over the `Create or connect a test account` button, check that the URL is `https://marketplace.stripe.com`:
    Screenshot_2024-08-05_at_2_26_59 AM
6. Click the `Create or connect a test account` button.
7. Login to Stripe (if needed)
8. Complete the Stripe App Oauth process, selecting an account, accepting the permissions and installing the app:
    image (2)
9. Check that the flow was successful by verifying the keys in the plugin settings:
    - Check that the `test_secret_key` starts with **`rk_`**
        354987227-859673ca-7f1b-4b9c-b779-5fc08425ac96-2
    - Check the `test_connection_type` was saved in the settings, with the value `app`:
        354987278-a1ca0376-f8dd-4db4-800d-a1e875e1c7ed-2

_NOTE_: The account selected in the Stripe App OAuth does not need to be the same one it was configured before... an account connected to our platform account can be selected; in this case, we'll still get the restricted keys and save the origin of the keys as `app`, but when generating the links for re-authentication (in the configure connection modal) we'll show the connect links as we check the controller to decide which flow to use.

---

-   [ ] Covered with tests (or have a good reason not to test in description ☝️)
-   [ ] Added changelog entry **in both** `changelog.txt` and `readme.txt` (or does not apply)
-   [ ] Tested on mobile (or does not apply)

**Post merge**

-   [ ] Added testing instructions to the [Release Testing Instructions wiki page](https://github.com/woocommerce/woocommerce-gateway-stripe/wiki/Release-Testing-Instructions) (or does not apply)
-   [ ] Added the [needs docs label](https://github.com/woocommerce/woocommerce-gateway-stripe/labels?q=docs) (or does not apply)
-   [ ] Included this PR in the Release Thread scope (or does not apply)

### CODE REVIEW

File: includes/connect/class-wc-stripe-connect-api.php [hunk 1]
```diff
@@ -71,13 +71,19 @@ public function get_stripe_oauth_init( $return_url, $mode = 'live' ) {
 		 * Send request to Connect Server for Stripe keys
 		 *
 		 * @param string $code OAuth server code.
+		 * @param string $type Optional. The type of the connection. 'connect' or 'app'. Default is 'connect'.
 		 * @param string $mode Optional. The mode to connect to. 'live' or 'test'. Default is 'live'.
 		 *
 		 * @return array
 		 */
-		public function get_stripe_oauth_keys( $code, $mode = 'live' ) {
+		public function get_stripe_oauth_keys( $code, $type = 'connect', $mode = 'live' ) {
```
Comment (james-allan) [inline includes/connect/class-wc-stripe-connect-api.php:79]: 📓 Just noting that adding a new param in the middle of the signature would typically be a bad idea but in this case the public version of `get_stripe_oauth_keys()` only has the `$code` param. ie `$mode` hasn't been released yet.

File: includes/connect/class-wc-stripe-connect.php [hunk 2]
```diff
@@ -66,27 +66,28 @@ public function get_oauth_url( $return_url = '', $mode = 'live' ) {
 		 *
 		 * @param string $state State token to prevent request forgery.
 		 * @param string $code  OAuth code.
+		 * @param string $type  Optional. The type of the connection. 'connect' or 'app'. Default is 'connect'.
 		 * @param string $mode  Optional. The mode to connect to. 'live' or 'test'. Default is 'live'.
 		 *
 		 * @return string|WP_Error
 		 */
-		public function connect_oauth( $state, $code, $mode = 'live' ) {
+		public function connect_oauth( $state, $code, $type = 'connect', $mode = 'live' ) {
```
Comment (james-allan) [inline includes/connect/class-wc-stripe-connect.php:74]: 📓 Same as above. The public version of `connect_oauth()` only has the `$state` and `$code` param. ie $mode hasn't been released yet.

File: includes/connect/class-wc-stripe-connect.php [hunk 3]
```diff
@@ -111,31 +112,36 @@ public function maybe_handle_redirect() {
...
+		 * @param stdObject $result OAuth's response result.
+		 * @param string    $type   Optional. The type of the connection. 'connect' or 'app'. Default is 'connect'.
 		 * @param string    $mode   Optional. The mode to connect to. 'live' or 'test'. Default is 'live'.
 		 *
-		 * @return stdObject|WP_Error OAuth response result or WP_Error.
+		 * @return stdObject|WP_Error OAuth's response result or WP_Error.
 		 */
-		private function save_stripe_keys( $result, $mode = 'live' ) {
-
+		private function save_stripe_keys( $result, $type = 'connect', $mode = 'live' ) {
```
Comment (james-allan) [inline includes/connect/class-wc-stripe-connect.php:136]: 📓 Same case here.

**Review by james-allan — APPROVED** [review]
Changes look good. I tested on `develop` and trying to connect via the app flow didn't work because when you returned to the store we couldn't fetch the keys. 

On this branch the keys are successfully fetched via the `/stripe/app-oauth-keys` endpoint. 

I also noted that the on this branch the `test_connection_type` was set to `app`.

On this branch I then also confirmed that I could still connect using the connect flow (`connect.stripe.com`) too and doing so still successfully fetched the keys and the `test_connection_type` was set to 'connect'. 

Lastly, even when the connection type is `app` the modal displays a Connected account status. 

Screenshot 2024-08-08 at 6 01 16 PM

File: includes/connect/class-wc-stripe-connect-api.php (hunk 1 above)
Comment (diegocurbelo) [reply includes/connect/class-wc-stripe-connect-api.php:79]: Yeah, I agree 👍🏼; as you mentioned below, the `mode` change is not public yet, so I chose to keep the same param order as in the connect server.

**Pull request merged**